oak/server/rust/oak_runtime/src/metrics/server.rs

```diff
@@ -0,0 +1,88 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use hyper::{
+    header::CONTENT_TYPE,
+    service::{make_service_fn, service_fn},
+    Body, Request, Response, Server,
+};
+
+use log::info;
+use prometheus::{Encoder, TextEncoder};
+use std::net::SocketAddr;
+
+#[derive(Debug)]
+enum MetricsServerError {
+    EncodingError(String),
+    ResponseError(String),
+}
+
+impl std::fmt::Display for MetricsServerError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            MetricsServerError::EncodingError(msg) => write!(f, "Metrics server error: {}", msg),
+            MetricsServerError::ResponseError(msg) => write!(f, "Metrics server error: {}", msg),
+        }
+    }
+}
+
+impl std::error::Error for MetricsServerError {}
+
+async fn serve_metrics(_req: Request<Body>) -> Result<Response<Body>, MetricsServerError> {
+    let encoder = TextEncoder::new();
+    let metric_families = prometheus::gather();
+    let mut buffer = vec![];
+    encoder.encode(&metric_families, &mut buffer).map_err(|e| {
+        MetricsServerError::EncodingError(format!("Could not encode metrics data: {}", e))
+    })?;
+
+    info!("Metrics size: {}", buffer.len());
+
+    Response::builder()
+        .status(http::StatusCode::OK)
+        .header(CONTENT_TYPE, encoder.format_type())
+        .body(Body::from(buffer))
+        .map_err(|e| {
+            MetricsServerError::ResponseError(format!("Could not build the response: {}", e))
```

Comment:
Does this end up being sent to the client? If so, we should not propagate details about internal errors, in general they may leak data. Better to log the error, and return safe, pre-defined error messages to the client. But maybe that's already the case, I don't know whether this `map_err` affects the response or not.

Reply:
I don't think so. According to the [documentation](https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err), `map_err` leave the `Ok` value untouched.